snuba/state/rate_limit.py

```diff
@@ -219,9 +232,18 @@ def __enter__(self) -> RateLimitStatsContainer:
         stats = RateLimitStatsContainer()
 
         for rate_limit_param in self.rate_limit_params:
-            child_stats = self.stack.enter_context(rate_limit(rate_limit_param))
-            if child_stats:
-                stats.add_stats(rate_limit_param.rate_limit_name, child_stats)
+            try:
+                child_stats = self.stack.enter_context(rate_limit(rate_limit_param))
+                if child_stats:
+                    stats.add_stats(rate_limit_param.rate_limit_name, child_stats)
+            except RateLimitExceeded as e:
+                # If an exception occurs in one of the rate limiters, the __exit__ callbacks are not
+                # called since the error happened in the __enter__ method and not in the context
+                # block itself. In the case that one of the rate limiters caught a limit, we need
+                # these exit functions to be called so we can roll back any limits that were set
+                # earlier in the stack.
+                self.__exit__(*sys.exc_info())
```

Comment:
It turned out that the fact that we were calling `rds.zincrby` in the `finally` clause meant that `rds.zincrby` was actually called after an exception. So technically yes, we were not exisitng the context but it did not cause any issue. SO no issue in adding this. Though did you add it because after the change you made above we were actually not calling `rds.zincrby` anymore ?

Reply:
zincrby was not being called in the failing `rate_limit` for sure, since the context was not being exited correctly. The main problem this is all fixing is that if an exception occurs in the `__enter__` method of a context, the `__exit__` method is not called. It needs to be manually invoked, as I did here.

Reply:
Did you actually test this behavior? I was expecting what you mention, but when I added a test to verify that, if the second rate limiter of the chain fail during __enter__, the first would not be exited. 
Though it did exit. I suspect it was because of the `finally` clause in which `zincrby` was called. 
I added logs around `zincrby` and, no matter whether the inner rate limiter succeeded or failed on __enter__ or failed, the external `zincrby` was always called.

Reply:
The _external_ one is called, but the _internal_ one is not called. So you're half right.

Reply:
But do you need the internal one to be called?
When the internal one raises, it already removed the key entirely, you do not want to try to decrement it.
https://github.com/getsentry/snuba/blob/master/snuba/state/rate_limit.py#L205-L219

Or is it now needed with this change?

Reply:
The documentation is pretty clear that its up to you to clean up yourself if the __enter__ fails. It's not expected that the __exit__ function is called. How did you test this? Because part of the problem is wrapping the `rate_limit` function in an ExitStack and then wrapping that in another context (RateLimitAggregator). The __enter__ method of `RateLimitAggregator` is called, and raises an exception (from `rate_limit.__enter__`). The __exit__ method of `RateLimitAggregator` is not called, which means the ExitStack is not closed and the yields are not called.

It's still a little confusing. If I comment this line out, then in my test, the second two test cases fail but the first one still passes.